contracts/mmr/MMRVerification.sol

```diff
@@ -0,0 +1,278 @@
+// "SPDX-License-Identifier: UNLICENSED"
+pragma solidity ^0.7.0;
+
+/**
+ *
+ * @dev MMRVerification library for MMR inclusion proofs generated
+ *      by https://github.com/nervosnetwork/merkle-mountain-range.
+ *
+ *      The index of this MMR implementation starts from 1 not 0.
+ */
+contract MMRVerification {
+
+    struct MountainData {
+        uint256 position;
+        bytes32 hash;
+        uint256 height;
+    }
+
+    mapping(uint256 => MountainData) public queue;
+
+    /**
+     * @dev Verify an MMR inclusion proof for a leaf at a given index.
+     */
+    function verifyInclusionProof(
+        bytes32 root,
+        bytes32 leafNodeHash,
+        uint256 leafIndex,
+        uint256 leafCount,
+        bytes32[] memory proofItems
+    )
+        public returns (bool)
+    {
+        // Input index must be a leaf
+        uint256 leafPos = leafIndexToPos(leafIndex);
+        if(!isLeaf(leafPos)) {
+            return false;
+        }
+
+        // Handle 1-leaf MMR
+        if(leafCount == 1 && leafPos == 1) {
+            return true;
+        }
+
+        // Calculate the index of our leaf's mountain peak
+        uint256 targetPeakIndex;
+        uint256 numLeftPeaks;
+        uint256[] memory peakIndexes = getPeakIndexes(leafCount);
+        for (uint256 i = 0; i < peakIndexes.length; i++) {
+            if (peakIndexes[i] >= leafPos) {
+                targetPeakIndex = peakIndexes[i];
+                break;
+            }
+            numLeftPeaks = numLeftPeaks + 1;
+        }
+
+        // Calculate our leaf's mountain peak hash
+        bytes32 mountainHash = calculatePeakRoot(
+            numLeftPeaks, leafNodeHash, leafPos, targetPeakIndex, proofItems
+        );
+
+        // All right peaks are rolled up into one hash. If there are any, bag them.
+        bytes32 bagger = mountainHash;
+        if(targetPeakIndex > leafPos) {
+            bagger = keccak256(abi.encodePacked(proofItems[proofItems.length-1], bagger));
```

Comment:
Wouldn't you need some `proofItems.length` checks to ensure it has enough entries? I guess accessing nonexistent entries will return `0` and then we would end up with incorrect proof?

Reply:
Yeah, nonexistent entries will result in an invalid proof - added a test case to demonstrate in https://github.com/Snowfork/verification-solidity/pull/8/commits/7141ee0c8f23e3f8755476fae5116bbbb183e509. Still, we could just return false for a proof that doesn't have any contents so I added a `proofItems.length` check at the top of the method as well.

Reply:
I removed the `proofItems.length` check, see https://github.com/Snowfork/verification-solidity/pull/8#discussion_r579680277.